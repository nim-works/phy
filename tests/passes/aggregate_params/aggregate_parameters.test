discard """
  description: "
    Ensure that aggregate parameters are turned into address (i.e., integer)
    parameters correctly, and that the argument expressions at the callsite
    are turned into an address-of operations accordingly
  "
  output: "(Done)"
"""
(TypeDefs
  (Record 4 4
    (Field 0 (Int 4)))
  (ProcTy (Type 0))
  (ProcTy (Void) (Type 0))
  (ProcTy (Void)))
(GlobalDefs)
(ProcDefs
  (ProcDef (Type 1)
    (Locals (Type 0))
    (List
      (Block (Params)
        (Return
          (Copy (Local 0))))))

  (ProcDef (Type 2)
    (Locals (Type 0) (Type 0))
    (List
      (Block
        (Params (Local 0))
          (Drop (Addr (Local 0)))
          (Asgn (Local 1)
            (Copy (Local 0)))
          (Asgn
            (Path (Int 4) (Local 0) 0)
            (IntVal 100))
          (Return))))

  (ProcDef (Type 3)
    (Locals (Type 0))
    (List
      (Block (Params)
        (Call (Proc 1)
          (Copy (Local 0)))
        (Call (Proc 1)
          (Load (Type 0)
            (Addr (Local 0))))
        (Call (Proc 1)
          (Call (Proc 0)))
        (Return)))))
