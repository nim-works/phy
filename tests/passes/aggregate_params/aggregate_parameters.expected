;$sexp
(TypeDefs
  (Record 4 4
    (Field 0 (Int 4)))
  (ProcTy (Void) (Ptr))
  (ProcTy (Void) (Ptr))
  (ProcTy (Void)))
(GlobalDefs)
(ProcDefs
  (ProcDef (Type 1)
    (Locals (Type 0) (Ptr))
    (List
      (Block
        (Params (Local 1))
        (Store (Type 0)
          (Copy (Local 1))
          (Copy (Local 0)))
        (Return))))
  (ProcDef (Type 2)
    (Locals (Ptr) (Type 0))
    (List
      (Block
        (Params (Local 0))
        (Drop
          (Copy (Local 0)))
        (Asgn (Local 1)
          (Load (Type 0)
            (Copy (Local 0))))
        (Asgn
          (Path (Int 4)
            (Deref (Type 0)
              (Copy
                (Local 0)))
            0)
          (IntVal 100))
        (Return))))
  (ProcDef (Type 3)
    (Locals (Type 0) (Type 0) (Type 0) (Type 0))
    (List
      (Block (Params)
        (Asgn (Local 1)
          (Copy (Local 0)))
        (Call (Proc 1)
          (Addr (Local 1)))
        (Asgn (Local 2)
          (Load (Type 0)
            (Addr (Local 0))))
        (Call (Proc 1)
          (Addr (Local 2)))
        (Call (Proc 0)
          (Addr (Local 3)))
        (Call (Proc 1)
          (Addr (Local 3)))
        (Return)))))
